proxy_ssh.go

```diff
@@ -0,0 +1,287 @@
+			}
+			sessions[c.RemoteAddr()]["client"] = client
+			return nil, err
+		},
+	}
+
+	conf.AddHostKey(private)
+
+	s.config = conf
+
+	s.callbackFn = func(c ssh.ConnMetadata) (*ssh.Client, error) {
+		meta, _ := sessions[c.RemoteAddr()]
+		s.logger.Infof("[prxy.ssh] %v", meta)
+		client := meta["client"].(*ssh.Client)
+		s.logger.Infof("[prxy.ssh] Connection accepted from: %s\n", c.RemoteAddr())
+		return client, nil
+	}
+	s.wrapFn = func(c ssh.ConnMetadata, r io.ReadCloser) (io.ReadCloser, error) {
+		s.reader = &ReadSession{
+			ReadCloser: r,
+			delimiter:  []byte("\n"),
+		}
+		return s.reader, nil
+	}
+	s.closeFn = func(c ssh.ConnMetadata) error {
+		s.logger.Infof("[prxy.ssh] Connection closed.")
+		return nil
+	}
+
+	return nil
+}
+
+func (s *SSHProxy) handle(conn net.Conn) error {
+	serverConn, chans, reqs, err := ssh.NewServerConn(conn, s.config)
+	defer conn.Close()
+	if err != nil {
+		s.logger.Error(errors.Wrap(interpreter("Failed to handshake", err), "ssh.prxy"))
+		return (err)
+	}
+
+	clientConn, err := s.callbackFn(serverConn)
+	defer clientConn.Close()
+	if err != nil {
+		s.logger.Error(errors.Wrap(err, "ssh.prxy"))
+		return (err)
+	}
+
+	go ssh.DiscardRequests(reqs)
+
+	for ch := range chans {
+
+		sshClientChan, clientReq, err := clientConn.OpenChannel(ch.ChannelType(), ch.ExtraData())
+		if err != nil {
+			s.logger.Error(errors.Wrap(interpreter(" Could not accept client channel: ", err), "ssh.prxy"))
+			return err
+		}
+
+		sshServerChan, serverReq, err := ch.Accept()
+		if err != nil {
+			s.logger.Error(errors.Wrap(interpreter(" Could not accept server channel: ", err), "ssh.prxy"))
+			return err
+		}
+
+		// Connect requests of ssh server and client
+		go func() {
+			s.logger.Debug("[prxy.ssh] Waiting for request")
+
+		r:
+			for {
+				var req *ssh.Request
+				var dst ssh.Channel
+
+				select {
+				case req = <-serverReq:
+					dst = sshClientChan
+				case req = <-clientReq:
+					dst = sshServerChan
+				}
+
+				// Check if connection is closed
+				if req == nil {
+					s.logger.Debug("[prxy.ssh] SSH Request is nil")
+					return
+				}
+
+				s.logger.Debugf("[prxy.ssh] Request: \n\n%s %s %s %s\n\n", dst, req.Type, req.WantReply, req.Payload)
+				b, err := dst.SendRequest(req.Type, req.WantReply, req.Payload)
+				if err != nil {
+					s.logger.Error(errors.Wrap(err, "ssh.prxy"))
+				}
+
+				if req.WantReply {
+					req.Reply(b, nil)
+				}
+
+				switch req.Type {
+				case "exit-status":
+					break r
+				case "exec":
+					s.logger.Debug("[prxy.ssh] SSH request 'EXEC' is not supported")
+				default:
+					s.logger.Debugf("[prxy.ssh] %s", req.Type)
+				}
+			}
+
+			sshServerChan.Close()
+			sshClientChan.Close()
+		}()
+
+		var wrappedServerChan io.ReadCloser = sshServerChan
+		var wrappedClientChan io.ReadCloser = sshClientChan
+
+		defer wrappedServerChan.Close()
+		defer wrappedClientChan.Close()
+
+		if s.wrapFn != nil {
+			wrappedClientChan, err = s.wrapFn(serverConn, sshClientChan)
```

Comment:
Instead of passing original channel I am passing custom struct with io.Reader implemented.